src/lib/consensus/constants.ml

```diff
@@ -1,98 +1,471 @@
-let slots_per_epoch = Unsigned.UInt32.(of_int (3 * c * k))
+  val to_time : t -> time
 
-module Slot = struct
-  let duration_ms = Int64.of_int block_window_duration_ms
+  val zero : t
+
+  val ( / ) : t -> t -> t
+
+  val ( * ) : t -> t -> t
 end
 
-module Epoch = struct
-  let size = slots_per_epoch
+module Constants_UInt32 :
+  M_intf
+  with type length = Length.t
+   and type time = Block_time.t
+   and type timespan = Block_time.Span.t = struct
+  type t = UInt32.t
+
+  type length = Length.t
+
+  type time = Block_time.t
+
+  type timespan = Block_time.Span.t
+
+  type bool_type = bool
+
+  let constant = UInt32.of_int
 
-  (** Amount of time in total for an epoch *)
-  let duration =
-    Time.Span.of_ms Int64.Infix.(Slot.duration_ms * int64_of_uint32 size)
+  let zero = UInt32.zero
+
+  let of_length = Fn.id
+
+  let to_length = Fn.id
+
+  let of_time = Fn.compose UInt32.of_int64 Block_time.to_int64
+
+  let to_time = Fn.compose Block_time.of_int64 UInt32.to_int64
+
+  let of_timespan = Fn.compose UInt32.of_int64 Block_time.Span.to_ms
+
+  let to_timespan = Fn.compose Block_time.Span.of_ms UInt32.to_int64
+
+  let ( / ) = UInt32.Infix.( / )
+
+  let ( * ) = UInt32.mul
 end
 
-module Checkpoint_window = struct
-  let per_year = 12
+module Constants_checked :
+  M_intf
+  with type length = Length.Checked.t
+   and type time = Block_time.Unpacked.var
+   and type timespan = Block_time.Span.Unpacked.var = struct
+  open Snarky_integer
+
+  type t = field Integer.t
+
+  type length = Length.Checked.t
+
+  type time = Block_time.Unpacked.var
+
+  type timespan = Block_time.Span.Unpacked.var
+
+  type bool_type = Boolean.var
 
-  let slots_per_year =
-    let one_year_ms =
-      Core.Time.Span.(to_ms (of_day 365.)) |> Float.to_int |> Int.to_int64
+  let constant c = Integer.constant ~m (Bignum_bigint.of_int c)
+
+  let zero = constant 0
+
+  let of_length = Length.Checked.to_integer
+
+  let to_length = Length.Checked.Unsafe.of_integer
+
+  let of_time = Fn.compose (Integer.of_bits ~m) Block_time.Unpacked.var_to_bits
+
+  let to_time = Fn.compose Block_time.Unpacked.var_of_bits (Integer.to_bits ~m)
+
+  let of_timespan =
+    Fn.compose (Integer.of_bits ~m) Block_time.Span.Unpacked.var_to_bits
+
+  let to_timespan =
+    Fn.compose Block_time.Span.Unpacked.var_of_bits (Integer.to_bits ~m)
+
+  let ( / ) (t : t) (t' : t) = Integer.div_mod ~m t t' |> fst
+
+  let ( * ) = Integer.mul ~m
+end
+
+let create' (type a b c)
+    (module M : M_intf
+      with type length = a
+       and type time = b
+       and type timespan = c)
+    ~(protocol_constants : (a, a, b) Genesis_constants.Protocol.Poly.t) :
+    (a, b, c) Poly.t =
+  let open M in
+  let c = constant Coda_compile_config.c in
+  let block_window_duration_ms =
+    constant Coda_compile_config.block_window_duration_ms
+  in
+  let k = of_length protocol_constants.k in
+  let delta = of_length protocol_constants.delta in
+  let sub_windows_per_window = c in
```

Comment:
As far as I know, the definition of `sub_windows_per_window` and `slots_per_sub_window` would be changed in the future once the constants of the network is fixed. By the time I created the sliding window #3916, I made a random choice to let them be `c` and `k`. I think the actual definition of `sub_windows_per_window` and `slots_per_sub_window` would depend on both `k` and the `f` (block production rate?). Maybe add a TODO here to remind us to change it before the mainnet.